feat: Migrate Website to Bootstrap 4

netchampfaris opened the pull request:
**New Features:**

- Bootstrap 4
- SEO
  - Meta tags will be added based on standard properties from the `context` object. Fields like `image`, `description`, `author` and `published_on` will be directly added as meta tags if they are found in the context object.
  - Introduce **Website Route Meta** DocType which can be used to add/override arbitrary meta tags on routes.
- Bootstrap 4 Theming
  - Generate new themes by overriding the standard Bootstrap 4 SASS variables. These themes will be generated on the fly.

<details>
<summary><b>Bootstrap 4 Theming Demo:</b></summary>
<img src="https://user-images.githubusercontent.com/9355208/53435333-92553b80-3a1e-11e9-87e8-feacf4919345.gif" />
</details>


<details>
<summary><b>Website Route Meta:</b></summary>
<img src="https://user-images.githubusercontent.com/9355208/53435101-122ed600-3a1e-11e9-9250-0e307e76b401.png" />
</details>

---

**Deprecated:**

- `js_globals`
No usage in any of our apps. You can directly write JS in `script` tags.
- `bootstrap` hook
Bootstrap 3 theming was by adding another stylesheet on top of it. This is why we had this hook. Now we theme by generating our own stylesheets in Website Theme.
- `/profile`
Profile page looked similar to the My Account page, added a redirect to `/me`

netchampfaris committed:
[fix] file test

netchampfaris committed:
feat: Migrate Website to Bootstrap 4

- Use bootstrap 4 base for website assets
- hooks.py: Remove bootstrap hook
- Upgrade navbar, breadcrumbs to 4
- Support multi-level dropdown

BREAKING CHANGE

netchampfaris committed:
Merge branch 'develop' of https://github.com/frappe/frappe into website-bootstrap-4

netchampfaris committed:
feat: Add metatags on every webpage

- Add keywords field in Web Page for SEO

netchampfaris committed:
Merge branch 'develop' of https://github.com/frappe/frappe into website-bootstrap-4

netchampfaris committed:
feat: Customize bootstrap using css variables (wip)

netchampfaris committed:
fix: Modal primary button

netchampfaris committed:
fix: styling for header actions

netchampfaris committed:
fix: Support disabled options in Select

netchampfaris committed:
fix: Remove ajax loader from msg_dialog

netchampfaris committed:
fix: styling

netchampfaris committed:
fix: Cleanup Blog, List, Comment templates for Bootstrap 4

- Bootstrap 4 class migration

netchampfaris committed:
fix: Indicator font size

netchampfaris committed:
fix: Style for standard image

netchampfaris committed:
wip: Auto generate css variables

netchampfaris committed:
fix: Urls starting with http are valid

netchampfaris committed:
fix: Urls starting with http are valid

netchampfaris committed:
fix: Replace toString check with a JSON check

- toString returns [object Object] which is not a very accurate check

netchampfaris committed:
fix: Override Dialog implementation for bootstrap 4

netchampfaris committed:
fix: Full width form layout

netchampfaris committed:
feat: Add get_single_value in jinja methods

netchampfaris committed:
fix: Cleanup navbar login

netchampfaris committed:
fix: Deprecate profile page

- Route to My Account page

netchampfaris committed:
feat: Generate Custom Bootstrap Theme using scss variables

netchampfaris committed:
fix: Add website_theme directory to .gitignore

netchampfaris committed:
fix: Refresh form after generating theme

netchampfaris committed:
fix: Auto generate css file on save

- Remove generate button

netchampfaris committed:
fix: Absolute css url

netchampfaris committed:
fix: Default white navbar

netchampfaris committed:
fix: Import popen

netchampfaris committed:
fix: Rename fields and cleanup

netchampfaris committed:
fix: Remove utility classes and style using css

netchampfaris committed:
fix: Add datepicker styles for webform

netchampfaris committed:
fix: Bootstrap 4 class fixes

netchampfaris committed:
fix: Add more meta tags from standard properties

netchampfaris committed:
Merge branch 'develop' of https://github.com/netchampfaris/frappe into website-bootstrap-4

netchampfaris committed:
style: Spaces to tabs

netchampfaris committed:
fix: Remove unused variables file

netchampfaris committed:
fix: Deprecate js_globals

netchampfaris committed:
fix: pass spacer to frappe.scrub

netchampfaris committed:
fix(Web Page): Remove old script style

- Set route from title

netchampfaris committed:
fix: Rename Theme Section properly

netchampfaris committed:
fix: Bootstrap default custom theme

netchampfaris committed:
fix: update web_page.json

netchampfaris committed:
Merge branch 'develop' of https://github.com/frappe/frappe into website-bootstrap-4

netchampfaris committed:
style: Missing semicolon

netchampfaris committed:
style: Unused variable

netchampfaris committed:
fix: Revert class name for Print Format

netchampfaris committed:
test: Update Blog Post test

netchampfaris committed:
fix: Revert class names for Print Format

netchampfaris committed:
style: ESLint

netchampfaris committed:
fix(pg): longtext is not valid in Postgres

rmehta commented on frappe/public/less/website.less:
```diff
@@ -420,12 +420,12 @@ a.active {
-	border-radius: 0px;
-	font-size: 12px;
-}
+// .page-breadcrumbs .breadcrumb {
```
remove, not comment

rmehta commented on frappe/templates/base.html:
```diff
@@ -3,101 +3,92 @@
 	<!-- js should be loaded in body! -->
-	<script type="text/javascript"
-		src="/assets/frappe/js/lib/jquery/jquery.min.js"></script>
-	<script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.0.4/socket.io.slim.js"></script>
```
woah! I promise I never saw this...

netchampfaris committed:
fix: Add get_link_doctype method in DocField

netchampfaris committed:
fix: Navbar dropdown items render

netchampfaris committed:
feat: Lazy load images only when they are in viewport

netchampfaris committed:
fix: Show label for Table fields

netchampfaris committed:
feat: Introduce Website Meta Tag

- Child Table to store meta tags with key and value
- Used in Web Page
- Remove redundant description field in Web Page
and patch it into Website Meta Tag

netchampfaris committed:
refactor: Website Meta Tag

Meta Tags are now decoupled from docs and are maintained per route basis
They will be tracked in Website Route Meta doctype where the `name`
is the `route`, it keeps track of a list of `key` and `value` in a table

Remove Meta Tag table from Web Page, add Set Meta Tag button helper.

netchampfaris committed:
test: Website Route Meta

netchampfaris committed:
style: Fix lint errors

netchampfaris commented:
@frappe-bot create instance

netchampfaris commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

Thunderbottom commented:
@frappe-bot create instance

netchampfaris commented:
@frappe-bot stop instance

netchampfaris commented:
@frappe-bot create instance ['erpnext#15965']

netchampfaris committed:
fix: Dont symlink node_modules if no public folder

netchampfaris committed:
fix: Update rollup and postcss plugin

netchampfaris committed:
fix: Common scss options for theme generator and rollup scss

netchampfaris committed:
fix: Make variables importable

netchampfaris committed:
fix: Put custom themes in /assets/css

netchampfaris committed:
fix: Don't include web_include_css when custom theme is set

Thunderbottom commented:
@frappe-bot stop instance

Thunderbottom commented:
@frappe-bot create instance ['erpnext#15965', 's-2vcpu-2gb']

scmmishra committed:
fix: Generate theme if new doc

Thunderbottom commented:
@frappe-bot stop instance

Thunderbottom commented:
@frappe-bot create instance ['erpnext#15965', 's-2vcpu-2gb']

Thunderbottom commented:
@frappe-bot stop instance

netchampfaris committed:
fix: Handle broken images

netchampfaris committed:
fix: Webpage meta patch

netchampfaris committed:
Merge branch 'develop' of https://github.com/frappe/frappe into website-bootstrap-4

netchampfaris committed:
fix(security): HTML escape all data- attributes

netchampfaris committed:
refactor: Website image classes

netchampfaris committed:
fix: Handle broken images with only CSS

- Remove JS handler for 404 images

netchampfaris committed:
fix(socketio): Return early if socketio not found

netchampfaris committed:
fix: UX for Website Route Meta

netchampfaris committed:
fix: Merge chat fields into one section

netchampfaris committed:
fix: Handle falsy homepage route

netchampfaris committed:
fix: Website Theme default value for custom theme

netchampfaris committed:
fix: Ignore TypeError while making thumbnail

netchampfaris committed:
fix: Empty header handler for slideshow template

netchampfaris committed:
fix: Website Theme custom naming

netchampfaris committed:
fix(Website Slideshow): Set header fieldtype as HTML Editor

netchampfaris committed:
fix(Website Theme): Remove Bootstrap link field

netchampfaris commented:
**ToDo**

Website Theme
- [ ] Remove Text Styles
- [ ] Remove Background Section

netchampfaris committed:
fix: Undo gitignore entry

netchampfaris committed:
fix: Import order

netchampfaris committed:
fix: Remove .pylintrc

netchampfaris committed:
style: Missing semicolon

netchampfaris committed:
fix: Update rollup-plugin-node-resolve

netchampfaris committed:
fix: Remove old style theming

- Remove Text Style section and fields
- Remove website_theme.css
- Add Configure Theme helper dialog to build a theme quickly

netchampfaris committed:
fix: Route check

netchampfaris committed:
Merge branch 'develop' of https://github.com/frappe/frappe into website-bootstrap-4

netchampfaris committed:
fix: Explicit enable flags

netchampfaris committed:
fix: Linting

netchampfaris committed:
fix: Comment test

netchampfaris committed:
Merge branch 'develop' of https://github.com/frappe/frappe into website-bootstrap-4

netchampfaris committed:
fix(snyk): Upgrade bootstrap and generate .snyk file

netchampfaris committed:
test: Uncomment table_multiselect test

netchampfaris committed:
tests: Wait for list to update

netchampfaris committed:
fix: Show slideshow indicator if more than 1 image

netchampfaris committed:
fix: Global Search fixes

- Update title, route and published in sync_value
- Update search_template for BS4

netchampfaris committed:
fix: Update slideshow for BS4

netchampfaris committed:
fix: Website Image css

scmmishra reviewed (approved):
All is well! 🎉
